### PR TITLE
arenadata_toolkit: add TOAST's size to adb_relation_storage_size

### PR DESCRIPTION
**Pull request by red1452**
This patch adds TOAST table's size to summary size of relation at C function
`adb_relation_storage_size`, which is used at SQL functions
`adb_relation_storage_size` and `adb_relation_storage_size_on_segments`.

The code is based on `calculate_toast_table_size` from dbsize.c, but without
calculating size of TOAST's indexes.

New test `adb_relation_storage_size_test` was added to check calculating sizes of
different tables (heap, ao, with TOAST or without) at SQL functions
`adb_relation_storage_size` and `adb_relation_storage_size_on_segments`.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/59704

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848384

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848385

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848386

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848379

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/849052

File: gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+CREATE TABLE ao_table_with_toast(a INT, b TEXT)
+WITH (APPENDOPTIMIZED=true, ORIENTATION=COLUMN)
+DISTRIBUTED BY (a);
+
+CREATE TABLE ao_table_without_toast(a INT, b INT)
+WITH (APPENDOPTIMIZED=true, ORIENTATION=COLUMN)
+DISTRIBUTED BY (a);
+
+-- Check that toast exist only for "with_toast" tables
+SELECT relname
```
Comment (RekGRpth) [inline gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql]:
```suggestion
SELECT relname, reltoastrelid != 0 with_toast
FROM pg_class
WHERE relname IN ('heap_table_with_toast', 'heap_table_without_toast', 'ao_table_with_toast', 'ao_table_without_toast');
```

File: gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      relnamespace IN (SELECT oid FROM pg_namespace WHERE nspname = 'public');
+
+-- Insert initial data to tables
+INSERT INTO heap_table_with_toast SELECT i, 'short_text' FROM generate_series(1,15) AS i;
+INSERT INTO heap_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;
+INSERT INTO ao_table_with_toast SELECT i, 'short_text' FROM generate_series(1,15) AS i;
+INSERT INTO ao_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;
+
+-- Check sizes on segments
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
```
Comment (RekGRpth) [inline gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql]:
```suggestion
WITH s AS (
    SELECT oid, relname
    FROM pg_class
    WHERE relname IN ('heap_table_with_toast', 'heap_table_without_toast', 'ao_table_with_toast', 'ao_table_without_toast')
) SELECT relname, size FROM s, arenadata_toolkit.adb_relation_storage_size_on_segments(oid);
```

File: gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    SELECT string_agg( chr(trunc(65+random()*26)::integer), '')
+    FROM generate_series(1,50000))
+WHERE a = 1;
+
+UPDATE ao_table_with_toast SET b = (
+    SELECT string_agg( chr(trunc(65+random()*26)::integer), '')
+    FROM generate_series(1,50000))
+WHERE a = 1;
+
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
```
Comment (RekGRpth) [inline gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql]:
```suggestion
WITH s AS (
    SELECT oid, relname
    FROM pg_class
    WHERE relname IN ('heap_table_with_toast', 'ao_table_with_toast')
) SELECT relname, size FROM s, arenadata_toolkit.adb_relation_storage_size_on_segments(oid);
```

File: gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'))
+ORDER BY 1;
+
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
+    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'))
+ORDER BY 1;
+
+-- Check summary size of tables
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
```
Comment (RekGRpth) [inline gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql]:
```suggestion
WITH s AS (
    SELECT oid, relname
    FROM pg_class
    WHERE relname IN ('heap_table_with_toast', 'heap_table_without_toast', 'ao_table_with_toast', 'ao_table_without_toast')
) SELECT relname, adb_relation_storage_size size FROM s, arenadata_toolkit.adb_relation_storage_size(oid);
```

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/59893

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858617

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858609

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858608

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858611

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858610

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/59918

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/860229

**Comment by bimboterminator1** [comment]
@red1452 , Ready to approve. Format your patch description with proper 80 aligment, please.

**Comment by red1452** [comment]
> @red1452 , Ready to approve. Format your patch description with proper 80 aligment, please.

Done.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/60148

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872166

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872167

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872164

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872160

File: gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+WHERE a = 1;
+SELECT relname, sizes.gp_segment_id, sizes.size
+FROM pg_class, arenadata_toolkit.adb_relation_storage_size_on_segments(oid) sizes
+WHERE relname IN ('heap_table_with_toast', 'ao_table_with_toast')
+ORDER BY 1, 2;
+        relname        | gp_segment_id | size  
+-----------------------+---------------+-------
+ ao_table_with_toast   |             0 |   160
+ ao_table_with_toast   |             1 | 50248
+ ao_table_with_toast   |             2 |   192
```
Comment (bimboterminator1) [inline gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]: @red1452 , Test with and without patch shows the same size of segment files for AO tables. And  the toast relation with `reltoastrelid` has 0 size.  Why is it happening?

Comment (red1452) [reply gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]: At this test, I add one row to toast table only at one segment. Empty toast table has zero size. Without patch sizes on segments will be the same as sizes before insert large value to this table.

Comment (bimboterminator1) [reply gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]: Without and with patch the size of all segment files is the same (50248 at first segment) after inserting large value.
And the result of `adb_relation_storage_size` is also the same.  Toast table has 0 size. `calculate_toast_table_size` returns zero size.

Comment (bimboterminator1) [reply gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]: Anyway, toast table has zero size for the column with large value. And test should show the case when toast table is filled eventually.

Comment (bimboterminator1) [reply gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]:
```
CREATE TABLE ao_table_with_toast(a INT, b TEXT)
WITH (APPENDOPTIMIZED=true, ORIENTATION=ROW);

insert into ao_table_with_toast select 1, (SELECT string_agg( chr(trunc(65+random()*26)::integer), '') FROM generate_series(1,50000));

SELECT oid::regclass AS Table,
       reltoastrelid::regclass As TOAST_Table,
       pg_relation_size(reltoastrelid) AS toast_size
FROM pg_class
WHERE relkind = 'r'
  AND reltoastrelid <> 0
ORDER BY 3 DESC; 
                  table                    |       toast_table       | toast_size 
--------------------------------------------+-------------------------+------------
ao_table_with_toast                        | pg_toast.pg_toast_16884 |      65536
```

Comment (bimboterminator1) [reply gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]: **ORIENTATION=ROW** works

Comment (red1452) [reply gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out]: You are right, fixed.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/60281

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879993

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879994

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879991

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879988

**Review by bimboterminator1 — DISMISSED** [review]

**Review by RekGRpth — DISMISSED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/60475

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/889983

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/889984

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/889981

**Review by RekGRpth — APPROVED** [review]

**Review by bimboterminator1 — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/60488

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890596

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890597

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890590

**Pull request merged**